File: stepik-union/src/main/java/com/jetbrains/tmp/learning/courseFormat/SectionNode.java
package com.jetbrains.tmp.learning.courseFormat;

import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import com.jetbrains.tmp.learning.core.EduNames;
import com.jetbrains.tmp.learning.stepik.StepikConnectorLogin;
import org.jetbrains.annotations.NotNull;
import org.stepik.api.client.StepikApiClient;
import org.stepik.api.exceptions.StepikClientException;
import org.stepik.api.objects.lessons.CompoundUnitLesson;
import org.stepik.api.objects.lessons.Lessons;
import org.stepik.api.objects.sections.Section;
import org.stepik.api.objects.sections.Sections;
import org.stepik.api.objects.units.Unit;
import org.stepik.api.objects.units.Units;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author meanmail
 */

public class SectionNode extends Node<Section, LessonNode, CompoundUnitLesson, StepNode> {
    private static final Logger logger = Logger.getInstance(SectionNode.class);

    public SectionNode() {
    }

    public SectionNode(@NotNull Project project, @NotNull Section data) {
        super(project, data);
    }

    @Override
    protected void loadData(long id) {
        try {
            StepikApiClient stepikApiClient = StepikConnectorLogin.authAndGetStepikApiClient();
            Sections sections = stepikApiClient.sections()
                    .get()
                    .id(id)
                    .execute();
            Section data;
            if (!sections.isEmpty()) {
                data = sections.getSections().get(0);
            } else {
                data = new Section();
                data.setId(id);
            }
            setData(data);
        } catch (StepikClientException logged) {
            logger.warn(String.format("Failed load section data id=%d", id), logged);
        }
    }

    @Override
    protected Class<LessonNode> getChildClass() {
        return LessonNode.class;
    }

    @Override
    public long getCourseId() {
        Section data = getData();
        return data != null ? getData().getCourse() : 0;
    }

    @Override
    protected Class<Section> getDataClass() {
        return Section.class;
    }

    @NotNull
    @Override
    String getDirectoryPrefix() {
        return EduNames.SECTION;
    }

    @Override
    protected List<CompoundUnitLesson> getChildDataList() {
        List<CompoundUnitLesson> objects = new ArrayList<>();
        try {
            StepikApiClient stepikApiClient = StepikConnectorLogin.getStepikApiClient();

            List<Long> unitsIds;
            Section data = getData();
            unitsIds = data != null ? data.getUnits() : Collections.emptyList();

            if (!unitsIds.isEmpty()) {
                Units units = stepikApiClient.units()
                        .get()
                        .id(unitsIds)
                        .execute();

                Map<Long, Unit> unitsMap = new HashMap<>();

                List<Long> lessonsIds = new ArrayList<>();

                units.getUnits().forEach(unit -> {
                    long lessonId = unit.getLesson();
                    lessonsIds.add(lessonId);
                    unitsMap.put(lessonId, unit);
                });

                Lessons lessons = stepikApiClient.lessons()
                        .get()
                        .id(lessonsIds)
                        .execute();

                lessons.getLessons()
                        .forEach(lesson -> objects.add(new CompoundUnitLesson(unitsMap.get(lesson.getId()), lesson)));
            }
        } catch (StepikClientException logged) {
            logger.warn("A section initialization don't is fully", logged);
        }

        return objects;
    }
}

File: stepik-core/src/main/java/org/stepik/core/courseFormat/SectionNode.kt
package org.stepik.core.courseFormat

import com.intellij.openapi.project.Project
import com.thoughtworks.xstream.annotations.XStreamOmitField
import org.stepik.api.client.StepikApiClient
import org.stepik.api.exceptions.StepikClientException
import org.stepik.api.objects.StudyObject
import org.stepik.api.objects.lessons.CompoundUnitLesson
import org.stepik.api.objects.sections.Section
import org.stepik.api.objects.units.Unit
import org.stepik.core.core.EduNames
import java.util.*


class SectionNode(project: Project? = null, stepikApiClient: StepikApiClient? = null, data: StudyObject? = null) :
        Node(project, stepikApiClient, data) {

    override val childClass: Class<LessonNode>
        get() = LessonNode::class.java

    override val childDataClass: Class<CompoundUnitLesson>
        get() = CompoundUnitLesson::class.java

    override val dataClass: Class<Section>
        get() = Section::class.java

    @XStreamOmitField
    override val directoryPrefix = EduNames.SECTION

    override fun loadData(stepikApiClient: StepikApiClient, id: Long): Boolean {
        try {
            val sections = stepikApiClient.sections()
                    .get()
                    .id(id)
                    .execute()

            val data: Section

            if (!sections.isEmpty) {
                data = sections.first
            } else {
                data = Section()
                data.id = id
            }

            val oldData = this.data as Section
            this.data = data
            return oldData.updateDate != data.updateDate
        } catch (logged: StepikClientException) {
            logger.warn(String.format("Failed load section data id=%d", id), logged)
        }

        return true
    }

    override fun getCourseId(stepikApiClient: StepikApiClient): Long {
        return (data as Section).course.toLong()
    }

    override fun getChildDataList(stepikApiClient: StepikApiClient): List<StudyObject> {
        val objects = ArrayList<CompoundUnitLesson>()
        try {
            val unitsIds: List<Long>
            val data = data as Section
            unitsIds = data.units

            if (!unitsIds.isEmpty()) {
                val size = unitsIds.size
                val list = unitsIds.toTypedArray()
                var start = 0
                var end: Int
                while (start < size) {
                    end = start + 20
                    if (end > size) {
                        end = size
                    }
                    val part = list.copyOfRange(start, end)
                    start = end
                    val units = stepikApiClient.units()
                            .get()
                            .id(*part)
                            .execute()

                    val unitsMap = HashMap<Long, Unit>()

                    val lessonsIds = ArrayList<Long>()

                    units.units.forEach { unit ->
                        val lessonId = unit.lesson.toLong()
                        lessonsIds.add(lessonId)
                        unitsMap[lessonId] = unit
                    }

                    val lessons = stepikApiClient.lessons()
                            .get()
                            .id(lessonsIds)
                            .execute()

                    lessons.lessons
                            .forEach { lesson ->
                                objects.add(CompoundUnitLesson(unitsMap[lesson.id],
                                        lesson))
                            }
                }
            }
        } catch (logged: StepikClientException) {
            logger.warn("A section initialization don't is fully", logged)
        }

        return objects
    }
}
